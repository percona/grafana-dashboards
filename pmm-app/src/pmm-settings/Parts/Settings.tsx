import { PluginTooltip, VerticalFormWrapper } from '../../react-plugins-deps/components/helpers/Helpers';
import { SelectField } from '../../react-plugins-deps/components/FieldsComponents/Select/Select';
import { ToggleField } from '../../react-plugins-deps/components/FieldsComponents/Toggle/Toggle';
import React, { ReactElement, useEffect, useState } from 'react';
import { Collapse, Slider } from 'antd';
import './Settings.scss';
import { InputField } from '../../react-plugins-deps/components/FieldsComponents/Input/Input';
import ButtonElement from '../../react-plugins-deps/components/FieldsComponents/Button/Button';
import { Form as FormFinal } from 'react-final-form';
import { useForm } from 'react-final-form-hooks';
import SettingsService from '../Settings.service';
import { showErrorNotification, showSuccessNotification } from '../../react-plugins-deps/components/helpers/notification-manager';
import { SliderField } from '../../react-plugins-deps/components/FieldsComponents/Slider/Slider';
import set = Reflect.set;

const { Panel } = Collapse;
const dataRetentionOptions = [{ value: 'h', label: 'Hours' }, { value: 'm', label: 'Minutes' }, { value: 's', label: 'Seconds' }];

const marks = {
  0: 'Low',
  1: 'Medium',
  2: 'High',
};

function callback(key) {
  console.log(key);
}

const customPanelStyle = {
  background: 'rgb(51, 51, 51)',
  marginBottom: 10,
  border: 1,
  borderColor: 'white',
  color: 'white',
  textColor: 'white',
};

const customCollapseStyle = {
  background: 'transparent',
  marginBottom: 10,
  border: 1,
  borderColor: 'black',
  color: 'white',
  textColor: 'white',
};

const dataRetentionValues = [
  {
    hr: '60s',
    mr: '180s',
    lr: '300s',
  },
  {
    hr: '5s',
    mr: '30s',
    lr: '300s',
  },
  {
    hr: '5s',
    mr: '5s',
    lr: '60s',
  },
];

const getMetricsResolutionValues = metricsResolutions => {
  if (metricsResolutions.hr == '5s' && metricsResolutions.mr == '5s' && metricsResolutions.lr == '60s') {
    return 2;
  } else if (metricsResolutions.hr == '5s' && metricsResolutions.mr == '30s' && metricsResolutions.lr == '300s') {
    return 1;
  } else if (metricsResolutions.hr == '60s' && metricsResolutions.mr == '180s' && metricsResolutions.lr == '300s') {
    return 0;
  } else {
    return 0;
  }
};
const SettingsPart = props => {
  const [loading, setLoading] = useState(false);
  const { settings } = props;
  const onSubmit = async values => {
    const settings = {
      data_retention: values.data_retention_count + values.data_retention_units,
      call_home: values.call_home,
      check_for_updates: values.check_for_updates,
      metrics_resolutions: dataRetentionValues[values.metrics_resolutions_slider],
    };
    setLoading(true);
    try {
      await SettingsService.setSettings({ settings: settings });
      setLoading(false);
      showSuccessNotification({ message: 'Settings updated' });
    } catch (e) {
      setLoading(false);
      showErrorNotification({ message: e.message });
    }
  };

  return (
    <FormFinal
      onSubmit={() => {}}
      render={(): ReactElement => {
        const { form, handleSubmit } = useForm({
          onSubmit: onSubmit,
          validate: () => {},
        });

        useEffect(() => {
          if (!settings.data_retention && !settings.metrics_resolutions) {
            return;
          }
          const [data_retention_count, data_retention_units] = [settings.data_retention.slice(0, -1), settings.data_retention.slice(-1)];
          let metrics_resolutions_slider = getMetricsResolutionValues(settings.metrics_resolutions);

          form.initialize(Object.assign(settings, { data_retention_count, data_retention_units, metrics_resolutions_slider }));
        }, [settings]);

        // @ts-ignore
        return (
          <form onSubmit={handleSubmit}>
            <>
              <VerticalFormWrapper
                label={'Metrics resolution'}
                tooltip={<PluginTooltip linkText={'Read more'} url={'#'} text={'This setting defines how frequently the data will be collected'} />}
                element={<SliderField marks={marks} form={form} defaultValue={2} name={'metrics_resolutions_slider'} />}
              />
              <Collapse bordered={false} defaultActiveKey={['1']} onChange={callback} style={customCollapseStyle}>
                <Panel header="Advanced settings " key="1" style={customPanelStyle}>
                  <VerticalFormWrapper
                    label={'Data retention'}
                    tooltip={<PluginTooltip linkText={'Read more'} url={'#'} text={'This is the value for how long data will be stored'} />}
                    element={
                      <div style={{ display: 'flex', justifyContent: 'flex-start', width: '100%' }}>
                        <InputField name={'data_retention_count'} form={form} wrapperStyle={{ width: '60%' }} />
                        <SelectField
                          form={form}
                          name={'data_retention_units'}
                          options={dataRetentionOptions}
                          defaultValue={'weeks'}
                          style={{ width: '40%' }}
                        />
                      </div>
                    }
                  />
                  <VerticalFormWrapper
                    label={'Call home'}
                    tooltip={
                      <PluginTooltip
                        linkText={'Read more'}
                        url={'#'}
                        text={'Option to send usage data back to Percona to let us make product better'}
                      />
                    }
                    element={<ToggleField form={form} name={'call_home'} />}
                  />
                  <VerticalFormWrapper
                    label={'Check for updates'}
                    tooltip={
                      <PluginTooltip linkText={'Read more'} url={'#'} text={'Option to check new versions and ability to update PMM from UI'} />
                    }
                    element={<ToggleField form={form} name={'check_for_updates'} />}
                  />
                </Panel>
              </Collapse>
              <ButtonElement onClick={() => {}} loading={loading} text={'Apply changes'} />
            </>
          </form>
        );
      }}
    />
  );
};

export default SettingsPart;
