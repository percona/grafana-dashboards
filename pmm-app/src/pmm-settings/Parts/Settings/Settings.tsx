import React, { ReactElement, useEffect, useState } from 'react';
import { Collapse } from 'antd';
import { Form as FormFinal } from 'react-final-form';
import { PluginTooltip } from '../../../react-plugins-deps/components/helpers/Helpers';
import SettingsService from '../../Settings.service';
import { showSuccessNotification } from '../../../react-plugins-deps/components/helpers/notification-manager';
import { SliderField } from '../../../react-plugins-deps/components/FormComponents/Slider/Slider';
import { ToggleField } from '../../../react-plugins-deps/components/FormComponents/Toggle/Toggle';
import { InputField } from '../../../react-plugins-deps/components/FormComponents/Input/Input';
import ButtonElement from '../../../react-plugins-deps/components/FormComponents/Button/Button';
import './Settings.scss';
import Validators from '../../../react-plugins-deps/components/validators/validators';
import { FormElement } from '../../../react-plugins-deps/components/FormComponents/FormElement/FormElement';
import { css } from 'emotion';
import { GUI_DOC_URL } from '../../panel.constants';

const { Panel } = Collapse;

const marks = {
  0: 'Low',
  1: 'Medium',
  2: 'High',
};

const style = {
  panel: css`
    background: #1f1d1d;
    margin-bottom: 10;
    border: none !important;
    color: white;
    text-color: white;
  `,
  collapse: css`
    background: transparent;
    margin-bottom: 10;
    border: none !important;
    color: white;
    text-color: white;
  `,
};

const dataRetentionValues: MetricsResolutionInterface[] = [
  {
    hr: '60s',
    mr: '180s',
    lr: '300s',
  },
  {
    hr: '5s',
    mr: '10s',
    lr: '60s',
  },
  {
    hr: '1s',
    mr: '5s',
    lr: '30s',
  },
];

export const getMetricsResolutionValues = metricsResolutions => {
  let metricsIndex = 0;

  dataRetentionValues.forEach((resolution, index) => {
    if (
      resolution.hr === metricsResolutions.hr &&
      resolution.mr === metricsResolutions.mr &&
      resolution.lr === metricsResolutions.lr
    ) {
      metricsIndex = index;
    }
  });

  return metricsIndex;
};

const SECONDS = 60;
const MINUTES = 60;
const HOURS = 24;
const SECONDS_IN_DAY = SECONDS * MINUTES * HOURS;
const MINUTES_IN_HOUR = MINUTES * HOURS;

const transformToDays = (count, units) => {
  switch (units) {
    case 'h':
      return count / HOURS;
    case 'm':
      return count / MINUTES_IN_HOUR;
    case 's':
      return count / SECONDS_IN_DAY;
    default:
      return '';
  }
};

interface MetricsResolutionInterface {
  lr: string;
  mr: string;
  hr: string;
}

interface SettingsInterface {
  data_retention: string;
  metrics_resolutions: MetricsResolutionInterface;
  disable_telemetry?: boolean;
  enable_telemetry?: boolean;
  disable_stt?: boolean;
  enable_stt?: boolean;
}

const SettingsPart = props => {
  const [loading, setLoading] = useState(false);
  const { settings } = props;
  const onSubmit = async values => {
    const updatedSettings: SettingsInterface = {
      // transform count in days to seconds
      data_retention: values.data_retention_count * SECONDS_IN_DAY + 's',
      metrics_resolutions: dataRetentionValues[values.metrics_resolutions_slider],
    };

    updatedSettings.disable_telemetry = !values.telemetry_enabled;
    updatedSettings.enable_telemetry = values.telemetry_enabled;
    updatedSettings.disable_stt = !values.stt_enabled;
    updatedSettings.enable_stt = values.stt_enabled;
    setLoading(true);
    try {
      await SettingsService.setSettings(updatedSettings);
      setLoading(false);
      showSuccessNotification({ message: 'Settings updated' });
    } catch (e) {
      setLoading(false);
    }
  };
  return (
    <FormFinal
      onSubmit={onSubmit}
      render={({ form, handleSubmit }): ReactElement => {
        useEffect(() => {
          if (!settings.data_retention && !settings.metrics_resolutions) {
            return;
          }
          const [count, units] = [settings.data_retention.slice(0, -1), settings.data_retention.slice(-1)];
          const sliderValue = getMetricsResolutionValues(settings.metrics_resolutions);

          form.initialize({
            ...settings,
            ...{
              data_retention_count: transformToDays(count, units),
              metrics_resolutions_slider: sliderValue,
              updates_disabled: !settings.updates_disabled,
            },
          });
        }, [settings]);
        // @ts-ignore
        return (
          <form onSubmit={handleSubmit}>
            <>
              <FormElement
                label="Metrics resolution"
                tooltip={
                  <PluginTooltip
                    links={[
                      {
                        url: `${GUI_DOC_URL}#server-admin-gui-metrics-resolution`,
                        text: 'Read more',
                      },
                    ]}
                    text="This setting defines how frequently the data will be collected"
                  />
                }
                element={
                  <SliderField
                    marks={marks}
                    defaultValue={2}
                    name="metrics_resolutions_slider"
                    tipFormatter={value => {
                      const values = dataRetentionValues[value];
                      return `high: ${values.hr}, medium: ${values.mr}, low: ${values.lr}`;
                    }}
                  />
                }
              />
              <Collapse
                bordered={false}
                defaultActiveKey={['1']}
                onChange={() => {}}
                className={style.collapse}
              >
                <Panel header="Advanced settings " key="1" className={style.panel}>
                  <FormElement
                    label="Data retention (In Days)"
                    tooltip={
                      <PluginTooltip
                        links={[
                          {
                            url:
                              'https://www.percona.com/doc/percona-monitoring-and-management/' +
                              '2.x/faq.html#how-to-control-data-retention-for-pmm',
                            text: 'Read more',
                          },
                        ]}
                        text="This is the value for how long data will be stored"
                      />
                    }
                    element={
                      <InputField
                        name="data_retention_count"
                        validate={Validators.compose(Validators.range(1, 3650), Validators.required)}
                        wrapperStyle={{ width: '100%' }}
                      />
                    }
                  />
                  <FormElement
                    label="Telemetry"
                    type="horizontal"
                    tooltip={
                      <PluginTooltip
                        links={[
                          {
                            url: `${GUI_DOC_URL}#server-admin-gui-telemetry`,
                            text: 'Read more',
                          },
                        ]}
                        text="Option to send usage data back to Percona to let us make product better"
                      />
                    }
                    element={
                      <ToggleField name="telemetry_enabled" disabled={form.getState().values.stt_enabled} />
                    }
                  />
                  <FormElement
                    label="Check for updates"
                    type="horizontal"
                    tooltip={
                      <PluginTooltip
                        links={[
                          {
                            url: `${GUI_DOC_URL}#check-for-updates`,
                            text: 'Read more',
                          },
                        ]}
                        text="Option to check new versions and ability to update PMM from UI"
                      />
                    }
                    element={<ToggleField name="updates_disabled" disabled />}
                  />
                  <FormElement
                    label="Security Threat Tool"
                    type="horizontal"
                    tooltip={
                      <PluginTooltip
                        links={[
                          {
                            url: `${GUI_DOC_URL}#security-threat-tool`,
                            text: 'Read more',
                          },
                        ]}
                        text="Enable Security Threat Tool and get updated checks from Percona"
                      />
                    }
                    element={
                      <ToggleField name="stt_enabled" disabled={!form.getState().values.telemetry_enabled} />
                    }
                  />
                </Panel>
              </Collapse>
              <ButtonElement loading={loading} text="Apply changes" />
            </>
          </form>
        );
      }}
    />
  );
};

export default SettingsPart;
