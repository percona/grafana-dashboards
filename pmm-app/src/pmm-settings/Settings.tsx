import React, {
  FC,
  useEffect,
  useMemo,
  useState
} from 'react';
import {
  Spinner,
  TabContent,
  Tab,
  useTheme
} from '@grafana/ui';
import { cx } from 'emotion';
import { TabsVertical } from 'shared/components/Elements/TabsVertical/TabsVertical';
import { Diagnostics } from './components/Diagnostics/Diagnostics';
import { MetricsResolution } from './components/MetricsResolution/MetricsResolution';
import { SSHKey } from './components/SSHKey/SSHKey';
import { AlertManager } from './components/AlertManager/AlertManager';
import { Advanced } from './components/Advanced/Advanced';
import { SettingsService, LoadingCallback } from './Settings.service';
import { Settings, TabKeys } from './Settings.types';
import { Messages } from './Settings.messages';
import { getSettingsStyles } from './Settings.styles';


export const SettingsPanel: FC = () => {
  const theme = useTheme();
  const styles = getSettingsStyles(theme);
  const [activeTab, setActiveTab] = useState('metrics');
  const {
    metrics,
    advanced,
    ssh,
    alertManager
  } = Messages.tabs;
  const tabs = useMemo(
    () => [
      { label: metrics, key: TabKeys.metrics, active: activeTab === TabKeys.metrics },
      { label: advanced, key: TabKeys.advanced, active: activeTab === TabKeys.advanced },
      { label: ssh, key: TabKeys.ssh, active: activeTab === TabKeys.ssh },
      { label: alertManager, key: TabKeys.alertManager, active: activeTab === TabKeys.alertManager },
    ],
    [activeTab]
  );
  const [settings, setSettings] = useState<Settings>();
  const [loading, setLoading] = useState(true);
  const updateSettings = async (body: any, callback: LoadingCallback) => {
    const response = await SettingsService.setSettings(body, callback);

    if (response) {
      setSettings(response);
    }
  };

  useEffect(() => {
    SettingsService.getSettings(setLoading, setSettings);
  }, []);

  return (
    <div className={styles.settingsWrapper}>
      <TabsVertical
        className={styles.tabsWrapper}
        dataQa="settings-tabs"
      >
        {tabs.map((tab, index) => (
          <Tab
            key={index}
            label={tab.label}
            active={tab.active}
            onChangeTab={() => setActiveTab(tab.key)}
          />
        ))}
      </TabsVertical>
      <TabContent
        className={cx(styles.tabContentWrapper, { [styles.settingsLoading]: loading })}
        data-qa="settings-tab-content"
      >
        {loading && <Spinner />}
        {(!loading && settings) && (
        <>
          {tabs[0].active
                && (
                <MetricsResolution
                  metricsResolutions={settings.metricsResolutions}
                  updateSettings={updateSettings}
                />
                )}
          {tabs[1].active
                && (
                <Advanced
                  dataRetention={settings.dataRetention}
                  telemetryEnabled={!!settings.telemetryEnabled}
                  updatesDisabled={!!settings.updatesDisabled}
                  sttEnabled={!!settings.sttEnabled}
                  updateSettings={updateSettings}
                />
                )}
          {tabs[2].active
                && (
                <SSHKey
                  sshKey={settings.sshKey || ''}
                  updateSettings={updateSettings}
                />
                )}
          {tabs[3].active
                && (
                <AlertManager
                  alertManagerUrl={settings.alertManagerUrl || ''}
                  alertManagerRules={settings.alertManagerRules || ''}
                  updateSettings={updateSettings}
                />
                )}
        </>
        )}
      </TabContent>
      <Diagnostics />
    </div>
  );
};
