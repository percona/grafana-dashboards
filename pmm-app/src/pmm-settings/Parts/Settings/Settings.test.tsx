import React from 'react';
import { shallow } from 'enzyme';
import Settings from './Settings';

jest.mock('shared/components/helpers/notification-manager');

// TODO: add correct shallow renderer, we need to have something to render to 2nd level
describe('Settings Part test', () => {
  it('Checkbox renders correct without props', () => {
    const root = shallow(
      <Settings
        settings={{
          data_retention: '24h',
          metrics_resolutions: {
            hr: '60s',
            mr: '180s',
            lr: '300s',
          },
        }}
      />,
    );

    expect(root).toMatchSnapshot();
    root.unmount();
  });
});
