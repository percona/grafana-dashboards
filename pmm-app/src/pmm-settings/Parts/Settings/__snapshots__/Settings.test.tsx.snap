// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings Part test Checkbox renders correct without props 1`] = `
<ReactFinalForm
  initialValues={
    Object {
      "metrics_resolutions_slider": 2,
    }
  }
  onSubmit={[Function]}
  render={[Function]}
/>
`;
